### ISSUE TITLE
Failure in git2 initialisation during use in rls

### ISSUE THREAD

**Issue by kinnison** [rust-lang/git2-rs#411]
Hi,

I have written a proc-macro which uses `git2-rs` (https://github.com/kinnison/git-testament) and it works just fine in the compiler.  But the rls seems unable to deal with it being present.  Instead it will crash under circumstances I have yet to fully determine, with the message:  `rls: libgit2/src/global.c:328: git__global_state: Assertion `git_atomic_get(&git__n_inits) > 0' failed.`

If I annotate `init()` at https://github.com/rust-lang/git2-rs/blob/master/libgit2-sys/lib.rs#L3033 then I am confident that the initialisation function is being called *somehow* yet it still explodes the rls with the above message.  This is easy to demonstrate by checking out git-testament, changing into the git tree, and running `rls --cli` and waiting.  It usually fails for me during the second or third thing built `cfg(test)` which usually is `testutils`.

Can you suggest a way for me to assist with diagnosing this issue?

Thanks,

Daniel.

**Comment by kinnison** [comment]
In order to proceed with the intended purpose of `git-testament` I've rewritten my proc-macro to shell out to `git` for the short term.  (I'd prefer to revert to `git2-rs` at some point) - to that end the repository has a branch `git2-version-before-rework` which can be used to investigate further.

**Comment by alexcrichton** [comment]
Sorry there's not a ton of bandwidth to help resolve these issues, but if possible it'd probably be best to open this up in a debugger and see which two stack traces call initialization, and from there we can probably debug further.

**Comment by kinnison** [comment]
I have become more confident in attaching debuggers to Rust code since I filed this issue so I'll go ahead and see if I can find out more.  Feel free to assign this issue to me for the record.

**Comment by kinnison** [comment]
I've managed to do a little research into this issue and what I've found is quite interesting/confusing.

The two times that the init function is called is when `cargo` initialises and when my proc-macro runs.

The two instances *seem* to, despite having different instances of libgit2 bound into them, shared the `git__n_inits` global in some sense since the return value from the init function the second time is 2.  However once we then call `git_repository_open_ext` the referenced `git__n_inits` ends up as zero and thus the assert fires.

Could this be a case of the symbols somehow being multi-exported from the cargo and proc-macro shared objects and thus being semi-badly unified?  The `git_libgit2_init()` in both cases is the same one (no debug symbols) but the `git__global_state()` call which fails has debug symbols, suggesting it's the one from my proc-macro built copy.

**Comment by kinnison** [comment]
I misspoke before, since cargo is linked into rls, however the rls binary does show an exported set of symbols for libgit2.

This might simply be a case of bad symbol namespacing where the libgit2-sys crate really ought not to be exporting the C symbols for libgit2 any further than itself.  Certainly `git_libgit2_init` is exported from both.  Interestingly the `git__n_inits` symbol is not present in either's symbol table, so I wonder exactly how the decision was made as to which `git_repository_open_ext` to call and why it didn't find the same variant as for `git_libgit2_init`

**Comment by alexcrichton** [comment]
Yeah I think you've hit the nail on the head here in diagnosing the issue.

I'm not really sure how to fix it  though because I thought the solution for these sorts of problems was to compile with `-fvisibility=hidden` which we're compiling libgit2 with. Other than that I'm not sure what to do with this :(

**Comment by kinnison** [comment]
According to some research I have done, that's not going to work given we then build a static library.  I think we need to end up passing `--exclude-libs path/to/libgit2.a` when we link it in, to transform the symbols from the static library into hidden ones.  We could pass `--exclude-libs ALL` but that seems a tad heavy handed given we can't know what other things might be linked in.  I shall give this a go and see if it hides the symbols from the proc-macro library I build.

**Comment by kinnison** [comment]
For reference, here's one of the things I found when googling around the topic: https://stackoverflow.com/questions/2222162/how-to-apply-fvisibility-option-to-symbols-in-static-libraries

**Comment by kinnison** [comment]
Okay I'm not sure I *can* get that passed into the link stage, do you know of a way?  You are wiser in the ins and outs of linking C code to Rust than I :grin:

**Comment by alexcrichton** [comment]
Hm reading up on `--exclude-libs` that looks like it's intended for being specified as an option after objects have been compiled whereas `-fvisibility=hidden` in theory should work because it's doing the same thing but at compile time instead of link time.

In any case there's unfortunately not a great way to test this out necessarily due to the usage of procedural macros, but you may be able to get away with `RUSTFLAGS=-Clink-args=-Wl,--exclude-libs,ALL` in the right location

**Comment by kinnison** [comment]
So far no banana :(  I think that needs to happen when the proc-macro is being linked into a `.so` but I'm not sure how to make that happen.
